report/web.py

```diff
@@ -50,32 +50,41 @@ class Benchmark:
   status: str
   result: run_one_experiment.AggregatedResult
   signature: str = ''
+  project: str = ''
+  function: str = ''
 
   def __post_init__(self):
-    self.signature = self.find_signature(self.id) or self.id
+    self.project = '-'.join(self.id.split('-')[1:-1])
+    self.function = self.id.split('-')[-1]
+    self.signature = self.find_signature() or self.id
 
-  @staticmethod
-  def find_signature(benchmark_id: str) -> str:
+  def find_signature(self) -> str:
     """
     Finds the function signature by searching for its |benchmark_id| in
     BENCHMARK_DIR.
     """
-    if not BENCHMARK_DIR:
+    project_path = os.path.join(BENCHMARK_DIR, f'{self.project}.yaml')
+    if not BENCHMARK_DIR or not os.path.isfile(project_path):
       return ''
 
-    for project_yaml in os.listdir(BENCHMARK_DIR):
-      yaml_project_name = project_yaml.removesuffix(".yaml")
-      with open(os.path.join(BENCHMARK_DIR, project_yaml)) as project_yaml_file:
-        if yaml_project_name not in benchmark_id:
-          continue
-        functions = yaml.safe_load(project_yaml_file).get('functions', [])
-        for function in functions:
-          function_name = benchmark_id.removeprefix(
-              f'output-{yaml_project_name}-')
-          if function.get('name', '').lower().startswith(function_name):
-            return f'{yaml_project_name}-{function.get("signature", "")}'
-
-    return ''
+    matched_prefix_signature = ''
+    with open(project_path) as project_yaml_file:
+      functions = yaml.safe_load(project_yaml_file).get('functions', [])
+      for function in functions:
+        function_name = function.get('name', '')
+        function_signature = function.get('signature', '')
+
+        # Best match is a full match, but sometimes the result directory only
+        # has the first n characters of a long function name so a full match is
+        # not possible.
+        # To avoid returning early on a prefix match when there is a full match
+        # farther down the list, we only return the prefix match at the end.
+        if function_name.lower() == self.function.lower():
+          return function_signature
+        if function_name.lower().startswith(self.function.lower()):
+          matched_prefix_signature = function_signature
```

Comment:
Thanks, this is clearer and more robust. 
1. Do you happen to notice any project/function only has a partial match? I don't recall which one it was.
2. Given now we handle this more carefully, shall we log a warning if there are multiple partial matches?

Reply:
https://github.com/google/oss-fuzz-gen/blob/main/benchmark-sets/all/ada-url.yaml

has both `ada_can_parse` and `ada_can_parse_with_base` . I only noticed it when I was looking for benchmarks with the same signature. These two were showing the same signature in the report and I tracked it down to this.

Reply:
A lot of the FI generated ones like immer are too long so the directories only have a substring.

I think this is the bit where long function names are truncated https://github.com/google/oss-fuzz-gen/blob/main/experiment/benchmark.py#L91

Reply:
Ah yes you are right.

Reply:
I think this is further motivation for #105

Reply:
Yep!